test/test_wave.py

```diff
@@ -20,4 +20,266 @@
+
+        nodes = discr.nodes().with_queue(queue)
+
+        # 2D: phi(x,y,t) = cos(omega*t-pi/4)*cos(x)*cos(y)
+        # 3D: phi(x,y,z,t) = cos(omega*t-pi/4)*cos(x)*cos(y)*cos(z)
+        # omega = sqrt(#dims)*c
+
+        c = 2.
+        offset = np.pi/4.
+        omega = np.sqrt(dim) * c
+
+        # 2D: u = phi_t = -omega*sin(omega*t-pi/4)*cos(x)*cos(y)
+        # 3D: u = phi_t = -omega*sin(omega*t-pi/4)*cos(x)*cos(y)*cos(z)
+        u = discr.empty(queue, dtype=np.float64)
+        u[:] = -omega * np.sin(-offset)
+        for i in range(dim):
+            u = u * clmath.cos(nodes[i])
+
+        # 2D: v[0] = c*phi_x = -c*cos(omega*t-pi/4)*sin(x)*cos(y)
+        #     v[1] = c*phi_y = -c*cos(omega*t-pi/4)*cos(x)*sin(y)
+        # 3D: v[0] = c*phi_x = -c*cos(omega*t-pi/4)*sin(x)*cos(y)*cos(z)
+        #     v[1] = c*phi_y = -c*cos(omega*t-pi/4)*cos(x)*sin(y)*cos(z)
+        #     v[2] = c*phi_z = -c*cos(omega*t-pi/4)*cos(x)*cos(y)*sin(z)
+        v = []
+        for i in range(dim):
+            v_i = discr.empty(queue, dtype=np.float64)
+            v_i[:] = -c * np.cos(-offset)
+            for j in range(dim):
+                if i == j:
+                    v_i = v_i * clmath.sin(nodes[j])
+                else:
+                    v_i = v_i * clmath.cos(nodes[j])
+            v.append(v_i)
+        fields = join_fields(u, v)
+
+        from mirgecom.wave import wave_operator
+        rhs = wave_operator(discr, c=c, w=fields)
+
+        # rhs(u part) = c*div(v)
+        # 2D: rhs(u part) = -2*c^2*cos(omega*t-pi/4)*cos(x)*cos(y)
+        # 3D: rhs(u part) = -3*c^2*cos(omega*t-pi/4)*cos(x)*cos(y)*cos(z)
+        expected_rhs_u = discr.empty(queue, dtype=np.float64)
+        # Why no negative sign?
+        # expected_rhs_u[:] = -omega**2 * np.cos(-offset)
+        expected_rhs_u[:] = omega**2 * np.cos(-offset)
+        for i in range(dim):
+            expected_rhs_u = expected_rhs_u * clmath.cos(nodes[i])
+
+        # rhs(v part) = c*grad(u)
+        # 2D: rhs(v part)[0] = omega*c*sin(omega*t-pi/4)*sin(x)*cos(y)
+        #     rhs(v part)[1] = omega*c*sin(omega*t-pi/4)*cos(x)*sin(y)
+        # 3D: rhs(v part)[0] = omega*c*sin(omega*t-pi/4)*sin(x)*cos(y)*cos(z)
+        #     rhs(v part)[1] = omega*c*sin(omega*t-pi/4)*cos(x)*sin(y)*cos(z)
+        #     rhs(v part)[2] = omega*c*sin(omega*t-pi/4)*cos(x)*cos(y)*sin(z)
+        expected_rhs_v = []
+        for i in range(dim):
+            rhs_v_i = discr.empty(queue, dtype=np.float64)
+            # Why negative sign?
+            # rhs_v_i[:] = omega*c * np.sin(-offset)
+            rhs_v_i[:] = -omega*c * np.sin(-offset)
+            for j in range(dim):
+                if i == j:
+                    rhs_v_i = rhs_v_i * clmath.sin(nodes[j])
+                else:
+                    rhs_v_i = rhs_v_i * clmath.cos(nodes[j])
+            expected_rhs_v.append(rhs_v_i)
+        expected_rhs = join_fields(expected_rhs_u, expected_rhs_v)
```

Comment:
I'd be in favor of trying to:

- get `sympy` (or even `pymbolic`) to generate the same math
- using their output (directly, without copy-pasting) to check the solution

so that the only hardcoded formula in here is the original solution.

Why is this useful (IMO)?

Right now, I would have to go and check all the calculus here by hand. That's maybe manageable here, but for more complex manufactured solutions, it'll get old pretty quick. So if we build up some machinery here to make this type of thing easier, I think that has the potential to pay off big time later on. I'm happy to guide you through how to do this, maybe Thursday.